Catch stray `{` in let-chains

sjwang05 opened the pull request:
Fixes #117766

sjwang05 committed:
Catch stray { in let-chains

sjwang05 committed:
Catch an edge case

sjwang05 committed:
Correctly handle while-let-chains

estebank commented on compiler/rustc_parse/src/lexer/tokentrees.rs:
```diff
@@ -121,9 +122,44 @@ impl<'a> TokenTreesReader<'a> {
             while parser.token != token::Eof {
                 if let Err(diff_err) = parser.err_diff_marker() {
                     diff_errs.push(diff_err);
+                } else if parser.is_keyword_ahead(0, &[kw::If, kw::While]) {
+                    in_cond = true;
+                } else if matches!(
+                    parser.token.kind,
+                    token::CloseDelim(Delimiter::Brace) | token::FatArrow
+                ) {
+                    // end of the `if`/`while` body, or the end of a `match` guard
+                    in_cond = false;
+                } else if in_cond && parser.token == token::OpenDelim(Delimiter::Brace) {
+                    // Store the `&&` and `let` to use their spans later when creating the diagnostic
+                    let maybe_andand = parser.look_ahead(1, |t| t.clone());
+                    let maybe_let = parser.look_ahead(2, |t| t.clone());
+                    if maybe_andand == token::OpenDelim(Delimiter::Brace) {
+                        // This might be the beginning of the `if`/`while` body (i.e., the end of the condition)
+                        in_cond = false;
+                    } else if maybe_andand == token::AndAnd && maybe_let.is_keyword(kw::Let) {
+                        let mut err = parser.struct_span_err(
+                            parser.token.span,
+                            "found a `{` in the middle of a let-chain",
+                        );
+                        err.span_suggestion(
+                            parser.token.span,
+                            "consider removing this brace to parse the `let` as part of the same chain",
+                            "", Applicability::MachineApplicable
+                        );
+                        err.span_note(
+                            maybe_andand.span.to(maybe_let.span),
+                            "you might have meant to continue the let-chain here",
+                        );
+                        errs.push(err);
+                    }
```
Let's move this logic to a separate method to make it less difficult to read the happy path. Either the new logic or the whole parser loop.

estebank commented on compiler/rustc_parse/src/lexer/tokentrees.rs:
```diff
@@ -121,9 +122,44 @@ impl<'a> TokenTreesReader<'a> {
+                            "consider removing this brace to parse the `let` as part of the same chain",
+                            "", Applicability::MachineApplicable
+                        );
+                        err.span_note(
```
```suggestion
                        err.span_label(
```

estebank reviewed (approved):
The approach is quite smart! Thank you for taking this on.

estebank commented on tests/ui/parser/brace-in-let-chain.stderr:
```diff
@@ -0,0 +1,71 @@
+error: this file contains an unclosed delimiter
+  --> $DIR/brace-in-let-chain.rs:58:54
+   |
+LL | fn main() {
+   |           - unclosed delimiter
+...
+LL | fn quux() {
+   |           - unclosed delimiter
+...
+LL | fn foobar() {
+   |             - unclosed delimiter
+...
+LL | fn fubar() {
+   |            - unclosed delimiter
+...
+LL | fn qux() {
+   |          - unclosed delimiter
+...
+LL | fn foo() {
+   |          - unclosed delimiter
+...
+LL | fn bar() {
+   |          - unclosed delimiter
+...
+LL | fn baz() {
+   |          - unclosed delimiter
+LL |     if false {
+LL |         {
+   |         - this delimiter might not be properly closed...
+LL |             && let () = ()
+LL |     }
+   |     - ...as it matches this but it has different indentation
+LL | }
+   |                                                      ^
```
I'm trying to decide if it's better in general to silence this error (like we do for diff markers) or not is best for these cases... I would leave it as is for now, because I don't know if I can come up with a case where doing so would be detrimental, but suspect we can silence it.

sjwang05 committed:
Move unclosed delim errors to separate function

sjwang05 commented:
Thanks for the feedback! I've moved all error handling logic into its separate method and replaced the `note` with a `label` as suggested.

> I'm trying to decide if it's better in general to silence this error (like we do for diff markers) or not is best for these cases

I just tried doing that, and strangely enough, all "found a `{` in the middle of a let-chain" errors except the first get silenced as well:
```
error: found a `{` in the middle of a let-chain
  --> $DIR/brace-in-let-chain.rs:6:24
   |
LL |         && let () = () {
   |                        ^
LL |         && let () = ()
   |         ------ you might have meant to continue the let-chain here
   |
help: consider removing this brace to parse the `let` as part of the same chain
   |
LL -         && let () = () {
LL +         && let () = ()
   |

error: aborting due to previous error
```
We should be getting 2 errors here, but we're only getting one 😕.

FWIW, here's what the silenced stderr would probably look like if I manage to get that problem sorted out:
```
error: found a `{` in the middle of a let-chain
  --> $DIR/brace-in-let-chain.rs:14:24
   |
LL |         && let () = () {
   |                        ^
LL |         && let () = ()
   |         ------ you might have meant to continue the let-chain here
   |
help: consider removing this brace to parse the `let` as part of the same chain
   |
LL -         && let () = () {
LL +         && let () = ()
   |

error: found a `{` in the middle of a let-chain
  --> $DIR/brace-in-let-chain.rs:6:24
   |
LL |         && let () = () {
   |                        ^
LL |         && let () = ()
   |         ------ you might have meant to continue the let-chain here
   |
help: consider removing this brace to parse the `let` as part of the same chain
   |
LL -         && let () = () {
LL +         && let () = ()
   |

error: aborting due to 2 previous errors
```
Though I feel like it's definitely easier to read, the unclosed delimiter errors aren't necessarily _wrong_ like they could be with diff markers, so silencing them might not be as helpful. The ideal solution is probably be something like the stderr you suggested in the original issue, where the label is printed inline with the unclosed delim error, though I have no clue how I'd go about doing that--the only way I can think of is modifying the errors in `report_suspicious_mismatch_block`, which would require somehow pairing up each erroneous let-chain with the correct `{`.

TaKO8Ki reviewed (commented):
Thank you for the great work! I left one suggestion.

sjwang05 committed:
Fix `is_keyword_ahead` visibility

Co-authored-by: Takayuki Maeda <takoyaki0316@gmail.com>

sjwang05 commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -1115,7 +1115,7 @@ impl<'a> Parser<'a> {
 
     /// Returns whether any of the given keywords are `dist` tokens ahead of the current one.
-    fn is_keyword_ahead(&self, dist: usize, kws: &[Symbol]) -> bool {
+    pub fn is_keyword_ahead(&self, dist: usize, kws: &[Symbol]) -> bool {
```
Fixed, thanks!

estebank commented:
@bors r=estebank,TaKO8Ki

bors commented:
:pushpin: Commit 274824b917c290d6ac9227e1c263a3358ba6e6f2 has been approved by `estebank,TaKO8Ki`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank,TaKO8Ki 274824b917c290d6ac9227e1c263a3358ba6e6f2 -->
<!-- homu: {"type":"Approved","sha":"274824b917c290d6ac9227e1c263a3358ba6e6f2","approver":"estebank,TaKO8Ki","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 274824b917c290d6ac9227e1c263a3358ba6e6f2 with merge ea1e5cc91f413224741011cbd3c2ed380d3d3129...
<!-- homu: {"type":"BuildStarted","head_sha":"274824b917c290d6ac9227e1c263a3358ba6e6f2","merge_sha":"ea1e5cc91f413224741011cbd3c2ed380d3d3129"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6844722328/job/18610874410)
Approved by: estebank,TaKO8Ki
Pushing ea1e5cc91f413224741011cbd3c2ed380d3d3129 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank,TaKO8Ki","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6844722328/job/18610874410"},"merge_sha":"ea1e5cc91f413224741011cbd3c2ed380d3d3129"} -->

rust-timer commented:
Finished benchmarking commit (ea1e5cc91f413224741011cbd3c2ed380d3d3129): [comparison URL](https://perf.rust-lang.org/compare.html?start=0828c159c58b2623d7750843a11f7bbd5e94ccb7&end=ea1e5cc91f413224741011cbd3c2ed380d3d3129&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0828c159c58b2623d7750843a11f7bbd5e94ccb7&end=ea1e5cc91f413224741011cbd3c2ed380d3d3129&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.8%, -0.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-0.8%, -0.3%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0828c159c58b2623d7750843a11f7bbd5e94ccb7&end=ea1e5cc91f413224741011cbd3c2ed380d3d3129&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.8%  | [3.8%, 3.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.7%  | [-0.4%, 3.8%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0828c159c58b2623d7750843a11f7bbd5e94ccb7&end=ea1e5cc91f413224741011cbd3c2ed380d3d3129&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.6%, 0.9%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-0.7%, 0.9%]  | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0828c159c58b2623d7750843a11f7bbd5e94ccb7&end=ea1e5cc91f413224741011cbd3c2ed380d3d3129&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.749s -> 674.078s (0.05%)
**Artifact size**: 311.08 MiB -> 311.07 MiB (-0.00%)
